module-system/sov-modules-api/src/transaction.rs

```diff
@@ -63,23 +96,181 @@ impl<C: Context> Transaction<C> {
 #[cfg(feature = "native")]
 impl<C: Context> Transaction<C> {
     /// New signed transaction.
-    pub fn new_signed_tx(priv_key: &C::PrivateKey, mut message: Vec<u8>, nonce: u64) -> Self {
+    pub fn new_signed_tx(
+        priv_key: &C::PrivateKey,
+        mut message: Vec<u8>,
+        chain_id: u64,
+        gas_tip: u64,
+        nonce: u64,
+    ) -> Self {
         // Since we own the message already, try to add the serialized nonce in-place.
         // This lets us avoid a copy if the message vec has at least 8 bytes of extra capacity.
-        let original_length = message.len();
-        message.extend_from_slice(&nonce.to_le_bytes());
+        let len = message.len();
+
+        // resizes once to avoid potential multiple realloc
+        message.resize(len + EXTEND_MESSAGE_LEN, 0);
+
+        message[len..len + 8].copy_from_slice(&chain_id.to_le_bytes());
+        message[len + 8..len + 16].copy_from_slice(&gas_tip.to_le_bytes());
+        message[len + 16..len + 24].copy_from_slice(&nonce.to_le_bytes());
 
         let pub_key = priv_key.pub_key();
         let signature = priv_key.sign(&message);
 
         // Don't forget to truncate the message back to its original length!
-        message.truncate(original_length);
+        message.truncate(len);
 
         Self {
             signature,
             runtime_msg: message,
             pub_key,
+            chain_id,
+            gas_tip,
             nonce,
         }
     }
 }
+
+impl<Tx> UnsignedTransaction<Tx> {
+    pub const fn new(tx: Tx, chain_id: u64, gas_tip: u64) -> Self {
+        Self {
+            tx,
+            chain_id,
+            gas_tip,
+        }
+    }
+}
+
+impl<Tx> borsh::BorshSerialize for UnsignedTransaction<Tx>
+where
+    Tx: borsh::BorshSerialize,
+{
+    fn serialize<W: io::Write>(&self, writer: &mut W) -> io::Result<()> {
+        self.tx.serialize(writer)?;
+        self.chain_id.serialize(writer)?;
+        self.gas_tip.serialize(writer)?;
+
+        Ok(())
+    }
+}
+
+impl<Tx> borsh::BorshDeserialize for UnsignedTransaction<Tx>
+where
+    Tx: borsh::BorshDeserialize,
+{
+    fn deserialize_reader<R: io::Read>(reader: &mut R) -> io::Result<Self> {
+        let tx = Tx::deserialize_reader(reader)?;
+        let chain_id = u64::deserialize_reader(reader)?;
+        let gas_tip = u64::deserialize_reader(reader)?;
+
+        Ok(Self {
+            tx,
+            chain_id,
+            gas_tip,
+        })
+    }
+}
+
+impl<Tx> serde::Serialize for UnsignedTransaction<Tx>
```

Comment:
Why do we need this implementation? Can's this be auto derived?

Reply:
In this case, we have one scenario where only `Serialize` is required

https://github.com/Sovereign-Labs/sovereign-sdk/blob/3ae9b7b92c8123864bf7471bfb4c086da93fe3af/module-system/sov-cli/src/workflows/transactions.rs#L108

I opted to avoid introducing new unnecessary bounds, as we can implement serde/borsh serialization conditionally. I can naturally just put the bounds downstream and get rid of the manual implementations, if they come to be a problem

Reply:
The manual implementation [here](https://github.com/Sovereign-Labs/sovereign-sdk/pull/1213/files/3ae9b7b92c8123864bf7471bfb4c086da93fe3af#diff-710780d3f43bf298801069e657ff34727958335421a0642a72a99c1d8b019e12R190) is quite complex.

The main issue lies in the hardcoded field names. If someone modifies or adds a new field, there's a risk of forgetting to update the code accordingly, which could lead to unnoticed errors. What are the downsides of adding additional bounds apart that there are nost strictly needed?

Reply:
Unfortunately that is a restriction of serde. They have no means to implement easily the serialization for such case.

We could require all the `Tx` instances to implement `{serde, borsh}::{Serialize, Deserialize}`, if they intend to use the `sov-cli`. I don't see much of a trouble by doing that, as it should be ok to have a over-restrictive CLI interface.

We can't, however, leak this to service level: as RPC, we might want only `borsh`. I'll make a shot to see if it can be done without leaking

Reply:
Done on https://github.com/Sovereign-Labs/sovereign-sdk/pull/1213/commits/aa3e742e72b17e54f3bb8c3d95e34324ebaaec09